Move rustdoc to 'How to build the compiler'

jyn514 opened the pull request:
The old layout was less about rustdoc and more about the compiler:
"Here is an example of a project using rustc_interface".
This gives focus to rustdoc on something you might want to work on for
its own sake. It also makes the page much easier to find for newcomers.

Addresses https://github.com/rust-lang/rustc-dev-guide/pull/750#issuecomment-645394137

Note that no other links need to be updated because the URL has not changed, only the links in the table of contents.

jyn514 commented:
Before: 
![image](https://user-images.githubusercontent.com/23638587/84915656-15bef400-b08b-11ea-947d-6ddda1098fec.png)


After: 
![image](https://user-images.githubusercontent.com/23638587/84915611-05a71480-b08b-11ea-83e4-7953715d8569.png)

mark-i-m commented:
Thanks @jyn514! Now that I'm looking at this, I'm actually not so sure... I don't feel that this chapter really fits in "Building and compiling" section... it's not really about that. Would it make sense to instead put the `rustdoc` chapter as 3.3 instead? That way it stands out a bit more, but it's still in a section about the compiler architecture.

jyn514 commented:
Well, `rustdoc.md` has sections about building and compiling rustdoc, it just also has sections about the rustdoc architecture in the same document. So I don't think it cleanly fits anywhere. The reason I don't think it fits in compiler architecture is because rustdoc is sort of separate from rustc, you don't need to understand it to understand the rest of the compiler. The reason I thought 'Building ...' was a good fit is because it's sort of an overview section.

JakobDegen commented:
To me it seems like that particular page is trying to do two things at once, one of which belongs where @jyn514 says, and the other where @mark-i-m says. In particular, it provides information about building/testing rustdoc, which seems out of place in Part 3 and imo needs to be moved somewhere more visible (evidenced by the fact that I wasn't able to find it), but also talks about the internal architecture of rustdoc which would be out of place in Part 1. Is there a reasonable way to split the page into two, or is this overkill?

jyn514 commented:
Splitting the page in two seems reasonable, I'll work on that when I have time.

mark-i-m commented:
@jyn514 Thanks! I think reorganizing the guide (especially the first part) will be priority soonish, so we may also see further action in the near future.

jyn514 commented:
I split up the two pages, fortunately there was a very nice break between the overview and the internals so it wasn't much work. I kept the overview in `1.2.5 Rustdoc` and moved the internals to `2.6 Rustdoc internals`: 
![image](https://user-images.githubusercontent.com/23638587/85467607-6cd83380-b579-11ea-8d25-ba8c78869c0e.png)

The overview now looks pretty sparse, but that's more because it's missing content than anything else.

jyn514 commented:
cc @GuillaumeGomez btw

mark-i-m commented:
Thanks @jyn514! This is great :)
